File: java/cekat-event-sdk-core/src/main/java/ai/cekat/events/error/CekatException.java
package ai.cekat.events.error;

/** Base class for every exception thrown by the Cekat SDK. */
public abstract class CekatException extends RuntimeException {
    private static final long serialVersionUID = 1L;

    /**
     * Creates an exception.
     *
     * @param message safe message that never contains the access token
     * @param cause underlying cause, or {@code null}
     */
    protected CekatException(String message, Throwable cause) {
        super(message, cause);
    }

    /**
     * Number of HTTP attempts made; zero when nothing was sent.
     *
     * @return attempts
     */
    public abstract int attempts();

    /**
     * Whether Cekat may have received the event even though no response was seen.
     *
     * @return {@code true} only for transport failures
     */
    public abstract boolean deliveryOutcomeUnknown();
}

File: dotnet/src/Cekat.EventSdk/Errors/CekatException.cs
namespace Cekat.EventSdk;

/// <summary>The base class for every SDK error. Messages never contain the access token.</summary>
public abstract class CekatException : Exception
{
    private protected CekatException(string message, int attempts, Exception? innerException = null)
        : base(message, innerException)
    {
        Attempts = attempts;
    }

    /// <summary>The number of HTTP attempts made. Zero for validation errors.</summary>
    public int Attempts { get; }

    /// <summary>
    /// <see langword="true"/> when no response was received, so Cekat may still have received the event
    /// and resending it can create a duplicate.
    /// </summary>
    public virtual bool DeliveryOutcomeUnknown => false;
}
